scripts/devnet.ts

```diff
@@ -0,0 +1,182 @@
+        console.error("");
+        console.error(msg);
+      } else {
+        console.error(err);
+      }
+      process.exit(1);
+    })
+    .wrap(Math.min(100, yargs.terminalWidth()))
+    .demandCommand().argv;
+}
+
+const upstreamCommand: yargs.CommandModule<
+  unknown,
+  { id: number; reset: boolean; bootstrap?: number; headless: boolean }
+> = {
+  command: "upstream <id>",
+  describe: "Run an upstream instance",
+  builder: yargs => {
+    return yargs
+      .positional("id", {
+        demandOption: true,
+        type: "number",
+        describe:
+          "Number identifying the instance. This is used to derive the peer configuration. Must be between 1 and 100",
+      })
+      .options({
+        bootstrap: {
+          type: "number",
+          describe: "Use the instance identified by the ID as a boostrap peer",
+        },
+        reset: {
+          type: "boolean",
+          default: false,
+          describe:
+            "Delete existing data for the peer and re-initialize the peer",
+        },
+        headless: {
+          type: "boolean",
+          default: false,
+          describe: "Only run radicle-proxy and not the frontend",
+        },
+      });
+  },
+  handler: async opts => {
+    const peerConfig = makePeerConfig(opts.id);
+    const missing = await Fs.access(peerConfig.radHome).catch(() => true);
+    if (opts.reset && !missing) {
+      await Fs.rm(peerConfig.radHome, { recursive: true });
+    }
+    if (missing || opts.reset) {
+      await execa(
+        "cargo",
+        ["run", "--bin", "radicle-proxy-init", "--", peerConfig.userHandle],
+        {
+          stdio: "inherit",
+          env: {
+            RAD_HOME: peerConfig.radHome,
+          },
+        }
+      );
+    }
+
+    let seedAddress: string | undefined;
+    if (opts.bootstrap !== undefined) {
+      seedAddress = getPeerAddress(makePeerConfig(opts.bootstrap));
+    }
+
+    if (opts.headless) {
+      await execa(
+        "cargo",
+        [
+          "run",
+          "--bin=radicle-proxy",
+          "--",
+          "--skip-remote-helper-install",
+          "--insecure-http-api",
+          "--unsafe-fast-keystore",
+          "--dev-log",
+        ],
+        {
+          stdio: "inherit",
+          env: getProxyEnv(peerConfig, seedAddress),
+        }
+      );
+    } else {
+      await execa("yarn", ["run", "electron", "./native/index.js"], {
+        stdio: "inherit",
+        env: {
+          NODE_ENV: "development",
+          RADICLE_UPSTREAM_UI_PROXY_ADDRESS: `127.0.0.1:${peerConfig.httpPort}`,
+          ...getProxyEnv(peerConfig, seedAddress),
+        },
+      });
+    }
+  },
+};
+
+main();
+
+interface PeerConfig {
+  userHandle: string;
+  peerId: string;
+  httpPort: number;
+  p2pPort: number;
+  radHome: string;
+}
+
+// Get a peer ID from a private key seed.
+//
+// Uses the same algorithm as `radicle-proxy-init`.
+function peerIdFromKeySeed(seed: string): string {
```

Comment:
This is pretty cool! Could be also useful for the p2p tests.

Reply:
Yes! Do you have an idea how we could leverage it?

Reply:
Hm, I looked at the p2p tests more closely, we're already using the output from `radicle-proxy-init` which outputs the peerId, so this wouldn't simplify anything.